### PR TITLE
Add BFloat16 support and use a new pack method for flash attention forward kernel

### PR DESCRIPTION
**Pull request by CaoE**
* Add BFloat16 support for BRGEMM flash attention forward kernel
* Use a new pack method instead of oneDNN pack for flash attention forward kernel to avoid the output leading dimension limitation.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/CPUBlas.cpp [hunk 1]
```diff
@@ -1172,6 +1175,10 @@ struct Pack : public KernelCache <PackKey, pack_t> {
     if (dtype == ScalarType::Half) {
       static bool fp16_pack = dnnl::get_effective_cpu_isa() >= dnnl::cpu_isa::avx512_core_amx_fp16;
       return fp16_pack;
+    } else if (dtype == ScalarType::BFloat16) {
+      // TODO Enable it on avx512_core_amx when bf16 perfomance is better
+      static bool bf16_pack = dnnl::get_effective_cpu_isa() >= dnnl::cpu_isa::avx512_core_amx_fp16;
```
Comment (jgong5) [inline aten/src/ATen/native/CPUBlas.cpp:1180]: why is it good with amx_fp16 then? Could you please explain in the note?

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 2]
```diff
@@ -291,6 +291,99 @@ inline void pad_remain_row_col_zero(
 
 }
 
+
+static inline void transpose_pad_2x32_block(
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:295]: Add function doc?

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 3]
```diff
@@ -291,6 +291,99 @@ inline void pad_remain_row_col_zero(
...
+    _mm512_storeu_si512(reinterpret_cast<__m512i*>(dst), d0);
+    _mm512_storeu_si512(reinterpret_cast<__m512i*>(dst + 32), d1);
+  }
+#else
+TORCH_CHECK(false, "transpose_pad_2x32_block is only supported when avx512 is supported")
+#endif
+}
+
+// If K % 2 != 0, pad K implicitly
+static inline void pack_vnni(
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:352]: ditto

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:298]: Added a brief description.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:358]: Added a brief description.

File: aten/src/ATen/native/CPUBlas.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp]: Enable packing on avx512_core_amx for bf16, and wait for the test results.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:358]: Also, is this implementation better?
https://github.com/intel/intel-extension-for-pytorch/blob/main/csrc/cpu/aten/kernels/WoqTppKrnl.cpp#L972

Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:358]: This implementation does not achieve performance improvements on bf16/fp16.

Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:358]: > Using pack_vnni2 is better? 

Currently only bf16/fp16 may use `pack_vnni`. If int8 and other types are considered in the future, can the vnni size be used as a template parameter to expand `pack_vnni ` ?

Comment (jgong5) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:358]: Yes, I mean for now we can name it `pack_vnni2` for clarity.

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@peterbell10 Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@peterbell10 Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
Hi @ezyang Could you please review this PR ? Thank you.

**Comment by ezyang** [comment]
I'm trying to understand if I need to do a deep review on this or not

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 4]
```diff
@@ -445,108 +525,76 @@ void cpu_flash_attention(
...
-    value_reorder_data = std::make_unique<scalar_t[]>(batchSize * num_head * kv_padding_size * rHeadSize);
-    value_reorder_ptr = value_reorder_data.get();
+    key_t_reorder = at::empty(
+      {batchSize, num_head, eheadSize, kvSize},
+      c10::CppTypeToScalarType<scalar_t>::value);
+    value_t_reorder = at::empty(
+      {batchSize, num_head, kv_padding_size, headSize},
+      c10::CppTypeToScalarType<scalar_t>::value);
+    key_reorder_ptr = key_t_reorder.data_ptr<scalar_t>();
+    value_reorder_ptr = value_t_reorder.data_ptr<scalar_t>();
```
Comment (peterbell10) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:544]: Probably has a bit more overhead this way, but sure. You could use `c10::DeviceArray` with the cpu allocator to get an allocation without the tensor overhead. Not a big deal though.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 5]
```diff
@@ -292,6 +292,105 @@ inline void pad_remain_row_col_zero(
...
+    uint16_t* dst,
+    int64_t ld_src,
+    int krem = 2,
+    int nrem = 32) {
+#if defined(CPU_CAPABILITY_AVX512)
+  __m512i r0, r1;
+  __m512i d0, d1;
+  // load
+  if (nrem < 32) {
+    __mmask32 mask_krem_v = (((long long)1) << nrem) - 1;
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp]: Thanks for your comments. Fixed.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 6]
```diff
@@ -292,6 +292,105 @@ inline void pad_remain_row_col_zero(
...
+  // store
+  if (nrem < 16) {
+    __mmask32 mask_rem_v = (((long long)1) << (nrem * 2)) - 1;
+    _mm512_mask_storeu_epi16(dst, mask_rem_v, d0);
+  } else if (nrem == 16) {
+    _mm512_storeu_si512(reinterpret_cast<__m512i*>(dst), d0);
+  } else if (nrem < 32) {
+    __mmask32 mask_rem_v = (((long long)1) << (nrem * 2 - 32)) - 1;
+    _mm512_mask_storeu_epi16(dst, mask_rem_v, d0);
+    _mm512_storeu_si512(reinterpret_cast<__m512i*>(dst), d0);
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp]: Thank you for reminding me, this is an omission and `_mm512_mask_storeu_epi16(dst, mask_rem_v, d0);` should be deleted.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 4 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:543]: Thank you very much for your suggestion. I have done some measurements before and found that the tensor creation overhead here has a negligible impact on the entire kernel.

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**